### PR TITLE
Implement promql meta endpoints

### PR DESCRIPTION
**Pull request by blagojts**
The LabelNames and LabelValues endpoints are required
so timescale-prometheus can be directly queried by Grafana.

Not all metadata endpoints are implemented as a
prometheus server has (the series  and metadata endpoints
are missing) but they are never queried by Grafana.

An http router is added to handle path params and scale
better than the default multiplexer under load

### CODE REVIEW

File: cmd/timescale-prometheus/main.go [hunk 1]
```diff
@@ -208,18 +208,44 @@ func main() {
...
+	labelsHandler := timeHandler(httpRequestDuration, "labels", api.Labels(queryable))
+	router.Get("/labels", labelsHandler)
+	router.Post("/labels", labelsHandler)
+	router.Get("/api/v1/labels", labelsHandler)
+	router.Post("/api/v1/labels", labelsHandler)
+
+	labelValuesHandler := timeHandler(httpRequestDuration, "label/:name/values", api.LabelValues(queryable))
+	router.Get("/label/:name/values", labelValuesHandler)
+	router.Get("/api/v1/label/:name/values", labelValuesHandler)
+
```
Comment (cevian) [inline cmd/timescale-prometheus/main.go:240]: I think we are missing a /api/v1/series endpoint https://prometheus.io/docs/prometheus/latest/querying/api/#finding-series-by-label-matchers

This can be in a separate PR

File: cmd/timescale-prometheus/main.go [hunk 2]
```diff
@@ -208,18 +208,44 @@ func main() {
...
 	writeHandler := timeHandler(httpRequestDuration, "write", api.Write(client, elector, &promMetrics))
-	http.Handle("/write", writeHandler)
-	http.Handle("/read", timeHandler(httpRequestDuration, "read", api.Read(client, &promMetrics)))
-	http.Handle("/query", queryHandler)
-	http.Handle("/query_range", queryRangeHandler)
-	http.Handle("/api/v1/query", queryHandler)
-	http.Handle("/api/v1/query_range", queryRangeHandler)
-	http.Handle("/healthz", api.Health(client))
+	router.Post("/write", writeHandler)
+
```
Comment (cevian) [inline cmd/timescale-prometheus/main.go:212]: I think we should document all the endpoint that we support. Doesn't have to be detailed, but a list/table of endpoints with links to the prom documentation would be good. Can be in a separate PR.

File: pkg/api/label_values.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		if err != nil {
+			respondError(w, http.StatusInternalServerError, err, "internal")
+			return
+		}
+		names, warnings, err := querier.LabelValues(name)
+		if err != nil {
+			respondError(w, http.StatusInternalServerError, err, "internal")
+			return
+		}
+
```
Comment (cevian) [inline pkg/api/label_values.go:34]: In prom, I see a code path for `w.WriteHeader(http.StatusNoContent)`  (api.go in the handler function). We don't have such a path. Should we (applies to all handlers)?

Comment (blagojts) [reply pkg/api/label_values.go:36]: Added a NoContent path in the respondQuery function, should cover the same path as in prometheus api impl

File: docs/prometheus_api.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+We implemented the endpoints for executing PromQL and labels metadata to allow the Timescale-Prometheus connector to be 
+used as a Prometheus compatible data-source in applications like Grafana. 
+
+<img src="./timescale-prometheus-arch.png" alt="Timescale-Prometheus Architecture Diagram" width="800"/>
+
+
+When requesting a query to run on data that is in the long-term storage through Prometheus, the query goes through the 
+following steps. 
+1. Query engine requests the data through the remote_read protocol;
```
Comment (cevian) [inline docs/prometheus_api.md:15]: can we add `1. The user issues a query to the prometheus server `

File: docs/prometheus_api.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+1. The connector translates the request to SQL statement with only a time range and label matchers applied;
+1. The database returns almost raw data;
+1. The connector marshals it into a format compatible with the remote_read protocol;
+1. The Query engine combines the local and remote data and applies any functions or aggregations before returning a 
+result.
+
+By having the Connector implement the PromQL APIs, the connector can:
+1. Parse the PromQL and translate it to a SQL statement that with a time range, label matchers, calculations and
```
Comment (cevian) [inline docs/prometheus_api.md:23]: please add `1. the user issues a query directly to the connector`

File: docs/prometheus_api.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+1. A modified Prometheus Query Engine finalizes the calculation and marshals the data to a proper response.
+
+By using the Connector for PromQL queries directly a network trip is avoided, and TimescaleDB is better utilized to 
+actually perform some calculations. 
+
+## Implemented Endpoints
+
+|                 Name               |              Endpoint                   |                          Description                      | 
+|------------------------------------|-----------------------------------------|-----------------------------------------------------------|
+|[Instant Queries][instant-queries]  |`GET /api/v1/query``POST /api/v1/query`  |Evaluate an instant query at a single point in time        |
```
Comment (cevian) [inline docs/prometheus_api.md:35]:
```suggestion
|[Instant Queries][instant-queries]  |`GET,POST  /api/v1/query`  |Evaluate an instant query at a single point in time        |
```
and similar below

File: pkg/api/labels.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	return parser.ValueTypeNone
+}
+
+func (l labelsValue) String() string {
+	return strings.Join(l, "\n")
+}
+
+func Labels(queriable *query.Queryable) http.Handler {
+	hf := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Access-Control-Allow-Origin", "*")
```
Comment (cevian) [inline pkg/api/labels.go:28]: This is controlled by a command line flag in Prom 
```
	a.Flag("web.cors.origin", `Regex for CORS origin. It is fully anchored. Example: 'https?://(domain1|domain2)\.com'`).
		Default(".*").StringVar(&cfg.corsRegexString)
```

Should we implement this in the future too (if so please file an issue)

Comment (cevian) [reply pkg/api/labels.go:28]: Also why isn't this in `setHeaders`?

File: pkg/pgmodel/end_to_end_tests/promql_endpoint_integration_test.go [hunk 8]
```diff
@@ -358,24 +510,9 @@ func testRequest(t *testing.T, req *http.Request, handler http.Handler, client *
 		}
 		defer tsResp.Body.Close()
 
-		var got, wanted response
-
-		err = json.Unmarshal(tsContent, &got)
+		err = comparator(promContent, tsContent)
```
Comment (cevian) [inline pkg/pgmodel/end_to_end_tests/promql_endpoint_integration_test.go:513]: should we also be comparing headers in this function? Not just the body? (ok to leave to another PR but let's file an issue if we do)

File: pkg/pgmodel/end_to_end_tests/promql_endpoint_integration_test.go [hunk 9]
```diff
@@ -320,20 +368,124 @@ func TestPromQLEndpoint(t *testing.T) {
...
+		if err != nil {
+			t.Fatalf("could not get label names from querier")
+		}
+		for _, label := range labelNames {
+			req, err = getLabelValuesRequest(apiURL, label)
+			if err != nil {
+				t.Fatalf("unable to create PromQL label values request: %v", err)
 			}
+			testMethod := testRequest(req, router, client, labelsResultComparator)
+			tester.Run(fmt.Sprintf("get label values for %s", label), testMethod)
```
Comment (cevian) [inline pkg/pgmodel/end_to_end_tests/promql_endpoint_integration_test.go:437]: let's also add a test for a request for a label name that does not exist

File: pkg/api/labels.go (hunk 7 above)
Comment (blagojts) [reply pkg/api/labels.go:28]: Moved to setHeaders, and created a card in project board

File: pkg/pgmodel/end_to_end_tests/promql_endpoint_integration_test.go (hunk 8 above)
Comment (blagojts) [reply pkg/pgmodel/end_to_end_tests/promql_endpoint_integration_test.go:513]: added to project board
